src/lib/plugins/ObjectStatePlugin/ObjectStateActions.ts

```diff
@@ -0,0 +1,111 @@
+import { uniqueId, filter } from 'lodash'
+import { Dispatch } from 'redux'
+import { Service, ServiceConstructor } from '../../core/Service'
+import { injectDispatch } from '../StorePlugin/StorePlugin'
+import { createSelectService, StateSelector } from '../StorePlugin/SelectService';
+import { getObjectType, ObjectTypeState } from './ObjectStatePlugin';
+
+export interface ObjectState<T> extends Service {
+  get(id: string): T
+  has(id: string): boolean
+  query(predicate: (x: T) => boolean): T[]
+  create(id: string, data: T): string
+  create(data: T): string
+  replace(id: string, data: T): void
+  patch(id: string, data: Partial<T>): void
+  delete(id: string): void
+}
+
+export function createObjectStateInterface<T>(objectType: string): ServiceConstructor<ObjectState<T>> {
+  class ObjectStateService extends Service implements ObjectState<T> {
+    @injectDispatch
+    private dispatch: Dispatch<{}>
+
+    @createSelectService(getObjectType(objectType))
+    private objects: StateSelector<ObjectTypeState<T>>
+
+    get(id: string): T {
+      const object = this.objects.value[id]
+      if (!object) {
+        throw new Error(`Could not find ${objectType} with id ${id}`)
+      }
+
+      return object
+    }
+
+    has(id: string): boolean {
+      return id in this.objects.value
+    }
+
+    query(predicate: (x: T) => boolean): T[] {
+      return filter(this.objects.value, (x) => Boolean(x && predicate(x))) as T[]
+    }
+
+    create(id: string, data: T): string
+    create(data: T): string
+    create(idOrData: any, data?: any) {
```

Comment:
Anything that can be done to break this up into sub functions that explain what it's doing a bit more clearly?